CPM port configuration for WTI OOB and PDU Devices

syncpak opened the pull request:
##### SUMMARY
Added the module cpm_portconfig to be able configure/query various port parameters of WTI OOB and PDU devices.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
cpm_portconfig

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```
ansible 2.8.0.dev0 (cpm_portconfig 1c18074346) last updated 2018/11/05 08:36:01 (GMT -700)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/kenp/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/kenp/ansible/lib/ansible
  executable location = /home/kenp/ansible/bin/ansible
  python version = 2.7.6 (default, Nov 23 2017, 15:49:48) [GCC 4.8.4]

```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Commit:
Add WTI OOB and PDU Device status, control and configuration module

Commit:
removed accidental file inclusions

Commit:
removed accidental file

Commit:
remove unneeded legacy files

Commit:
Merge remote-tracking branch 'upstream/devel' into devel

Commit:
Merge remote-tracking branch 'upstream/devel' into devel

Commit:
Merge remote-tracking branch 'upstream/devel' into devel

Commit:
Merge remote-tracking branch 'upstream/devel' into devel

Commit:
added cpm_portconfig to Configure WTI OOB and PDU device serial ports

Commit:
documentation was not updated with new parameter names

Commit:
EXAMPLES documentation updated

Commit:
changed exmaple parameters

mattsene commented:
Does this only include serial ports and not network ports ?

syncpak commented:
@mattsene This module is only for the serial ports of the WTI DSM/VMU units. The Network module is being worked on as we speak.

actionemb commented:
Seems to work. We have a DSM and a CPM and this modules seems to works.

syncpak commented:
shipit

syncpak commented:
shipit

mattsene commented:
shipit

vientiane commented:
shipit

wtinetworkgear commented:
shipit

scarshouldofwon commented:
shipit

syncpak commented:
Hi @gundalow thanks for merging our other two modules:
https://github.com/ansible/ansible/pull/47989
https://github.com/ansible/ansible/pull/47683

Is there anything special or problems on this one to get it merged also ?
If we goofed up, please let me know.
Thanks

gundalow commented on lib/ansible/modules/remote_management/cpm/cpm_portconfig.py:
```diff
@@ -0,0 +1,304 @@
+        cpm_url=dict(type='str', required=True),
+        cpm_username=dict(type='str', required=True),
+        cpm_password=dict(type='str', required=True, no_log=True),
+        serial_port=dict(type='str', required=True),
```
```suggestion
        serial_port=dict(type='int', required=True),
```

gundalow committed:
Update lib/ansible/modules/remote_management/cpm/cpm_portconfig.py

Co-Authored-By: syncpak <41302157+syncpak@users.noreply.github.com>

syncpak commented:
shipit

samdoran commented on lib/ansible/modules/remote_management/cpm/cpm_portconfig.py:
```diff
@@ -0,0 +1,304 @@
+    module = AnsibleModule(argument_spec=module_args, supports_check_mode=True)
+
+    if module.check_mode:
+        return result
```
This module fails in check mode since it is not calling `module.exit_json()`. Since this module does not look at the state and compare it before making changes, I would suggest removing check mode support from the module currently.

samdoran commented on lib/ansible/modules/remote_management/cpm/cpm_portconfig.py:
```diff
@@ -0,0 +1,304 @@
+        cpm_url=dict(type='str', required=True),
+        cpm_username=dict(type='str', required=True),
+        cpm_password=dict(type='str', required=True, no_log=True),
+        serial_port=dict(type='int', required=True),
```
I think all the `serial_` prefixes are repetitive. I think this module should be renamed `cpm_serial_port` then you can just remove all the `serial_` prefixes. This would also remove confusion on whether this module is configuring serial or network ports.

samdoran reviewed (changes requested):
This module does work in my testing. Here are the things I would like to see changed:
- rename this to `cpm_serial_port` or `cpm_serial_port_config`
- Remove the `action` parameter and make this two separate modules: one for setting the config, the other for getting info (`cpm_serial_port_info` or `cpm_port_info` for getting info about all ports)
- remove the `serial_` prefix from the parameter names
- This module always reports a change. It should evaluate the state of the device compared to the desired state from the playbook, then make/report changes appropriately.

On a side note, we have a [module defaults](https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html?highlight=module_defaults) section in playbooks now so modules can share common options. I would recommend creating a `cpm` group in order to make it easy to share the common settings across modules. The module groups are configured in [module_defaults.yml](https://github.com/ansible/ansible/blob/devel/lib/ansible/config/module_defaults.yml.)

syncpak commented:
Thanks @samdoran , in one day when our service department tests the new modules were we split up the modules into a cpm_serial_port_config and cpm_serial_port_info, removed the action parameter, trimmed the "serial_" off parameter names and properly use the change flag by querying the WTI device first. We will close this branch and open up two for the config and info modules for our serial ports.  @wtinetworkgear, @actionemb , @mattsene  changes will be coming soon for the new modules, we will post the new module locations.

samdoran commented:
Sounds good. I look forward to reviewing them.

syncpak commented:
Hi All, this module has been replaced by the module cpm_serial_port_config https://github.com/ansible/ansible/pull/55102 
When the initial round of reviews goes positive we will add the info serial port module that gets the parameters from the ports.
